lib/Command/RepairNotifications.php

```diff
@@ -0,0 +1,115 @@
+ * as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License, version 3,
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>
+ *
+ */
+
+namespace OCA\Notifications\Command;
+
+use OCP\IDBConnection;
+use Symfony\Component\Console\Command\Command;
+use Symfony\Component\Console\Input\InputArgument;
+use Symfony\Component\Console\Input\InputInterface;
+use Symfony\Component\Console\Output\OutputInterface;
+
+class RepairNotifications extends Command {
+
+	/** @var IDBConnection */
+	protected $connection;
+
+	public static $availableSubjects = [
+		'relativeLinks'
+	];
+
+	/**
+	 * @param IDBConnection $connection
+	 */
+	public function __construct(IDBConnection $connection) {
+		parent::__construct();
+		$this->connection = $connection;
+	}
+
+	protected function configure() {
+		$this
+			->setName('notifications:repairNotifications')
+			->setDescription('Repair existing notifications')
+			->addArgument('subject', InputArgument::REQUIRED, 'Subject to repair')
+		;
+	}
+
+	/**
+	 * @param InputInterface $input
+	 * @param OutputInterface $output
+	 * @return false|int
+	 */
+	protected function execute(InputInterface $input, OutputInterface $output) {
+		$subject = $input->getArgument('subject');
+
+		if (!\in_array($subject, self::$availableSubjects)) {
+			throw new \LogicException('Invalid subject');
+		}
+
+		$sql = $this->connection->getQueryBuilder();
+		$sql->select(['notification_id', 'link', 'actions'])
+			->from('notifications')
+			->where($sql->expr()->like('link', $sql->createPositionalParameter('http%')))
+			->orWhere($sql->expr()->like('actions', $sql->createPositionalParameter('%"link":"http%')));
+
+		$result = $sql->execute()->fetchAll();
+
+		if (!$result) {
+			$output->writeln('No notifications found to repair.');
+			return 0;
+		}
+
+		$output->writeln(\sprintf('%s notification(s) found to repair', \count($result)));
+
+		foreach ($result as $row) {
+			$output->writeln(\sprintf('Repairing notification with ID %s...', $row['notification_id']));
+
+			$sql = $this->connection->getQueryBuilder();
+			$sql->update('notifications')
+				->where($sql->expr()->eq('notification_id', $sql->createNamedParameter($row['notification_id'])));
+
+			$linkUrlComponents = \parse_url($row['link']);
+			if (\array_key_exists('scheme', $linkUrlComponents)) {
+				$newLink = \parse_url($row['link'], PHP_URL_PATH);
+				$sql->set('link', $sql->createNamedParameter($newLink));
+			}
+
+			if (\strpos($row['actions'], 'http') !== false) {
+				$actions = \json_decode($row['actions'], true);
+
+				foreach ($actions as &$action) {
```

Comment:
The `foreach (... as &$action)` has some caveats (using a reference for the item), so I'd rather use something different to avoid future problems.